### PR TITLE
Move `Rack::Session` to separate gem.

### PR DESCRIPTION
**Pull request by ioquatix**
This includes https://github.com/rack/rack/pull/1177 + moving code to `rack-session` gem.

### CODE REVIEW

**Comment by ioquatix** [comment]
I've moved all the code, and added references to the repo here so that the tests can pass, confirming the successful migration. Ignore the external tests, as they aren't able to pull in the right source gems.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
rack/encryptor should be put in the new rack-session gem, and not included with rack. Unless you have plans to use it elsewhere in rack?

Do we plan on keeping the gem dependency on rack-session when rack 3 is released?

**Comment by ioquatix** [comment]
> rack/encryptor should be put in the new rack-session gem, and not included with rack.

Sorry, this was my mistake, it should be deleted in the subsequent commit, but I missed it.

> Do we plan on keeping the gem dependency on rack-session when rack 3 is released?

I think that's a good question and I don't have a strong opinion, but I guess I could lean slightly towards no. Obviously the compatibility would be a bit more tricky with this direction but it allows us to move forward more cleanly.

**Comment by tenderlove** [comment]
> rack/encryptor should be put in the new rack-session gem, and not included with rack. Unless you have plans to use it elsewhere in rack?

+1

> Do we plan on keeping the gem dependency on rack-session when rack 3 is released?

I don't think we should.

**Comment by ioquatix** [comment]
Okay:

- Removed `rack-session` from `rack.gemspec` seems we all agree on that.
- Fixed incorrectly added `lib/rack/encryptor.rb`.

**Review by jeremyevans — APPROVED** [review]
Looks good, but I would remove the entry from the Gemfile (I assume it isn't needed).

File: Gemfile [hunk 1]
```diff
@@ -4,6 +4,9 @@ source 'https://rubygems.org'
 
 gemspec
 
+# Temporary to pull in unreleased gem:
+gem "rack-session", git: "https://github.com/rack/rack-session.git"
```
[the remark this answers is not in the document]
Comment (ioquatix) [reply Gemfile:8]: Yes, now the gem is released, we can move it to the gemspec.

**Comment by ioquatix** [comment]
@tenderlove I've added `rack-session` as a dependency for Rack (3.0). I suggest we remove this in Rack 3.1. It looks like Rack 3.0 will have a bit of compatibility shims to make the migration a bit easier. It will mean Rack 2.x and 3.x compatible code can work for a while and Rack 3.1 will then drop `rack-session`. Apps that want to pull in `rack-session` can do so and work with Rack 3.0 or 3.1 more easily.

**Comment by matthewd** [comment]
Dropping a dependency in a non-major version feels more likely to create surprises to me. If rack-session is not compatible with Rack 2.x, can we fix that? (Either within the gem, or by having future versions depend upon Rack >= 3, and releasing a zero-file rack-session 2.0 that depends on Rack < 3?)

**Comment by ioquatix** [comment]
@matthewd I don't have a strong opinion so feel free to propose a better approach which I hope you can also assist with implementation?

**Pull request merged**